src/Url.php

```diff
@@ -0,0 +1,243 @@
+<?php
+
+namespace Graby\HttpClient\Plugin\ServerSideRequestForgeryProtection;
+
+use Graby\HttpClient\Plugin\ServerSideRequestForgeryProtection\Exception\InvalidURLException;
+use Graby\HttpClient\Plugin\ServerSideRequestForgeryProtection\Exception\InvalidURLException\InvalidDomainException;
+use Graby\HttpClient\Plugin\ServerSideRequestForgeryProtection\Exception\InvalidURLException\InvalidIPException;
+use Graby\HttpClient\Plugin\ServerSideRequestForgeryProtection\Exception\InvalidURLException\InvalidPortException;
+use Graby\HttpClient\Plugin\ServerSideRequestForgeryProtection\Exception\InvalidURLException\InvalidSchemeException;
+
+class Url
+{
+    /**
+     * Validates the whole URL.
+     *
+     * @param string $url
+     * @param Options $options
+     *
+     * @throws InvalidURLException
+     *
+     * @return array
+     */
+    public static function validateUrl($url, Options $options)
+    {
+        if ('' === trim($url)) {
+            throw new InvalidURLException('Provided URL "'.$url.'" cannot be empty');
+        }
+
+        //Split URL into parts first
+        $parts = parse_url($url);
+
+        if (empty($parts)) {
+            throw new InvalidURLException('Error parsing URL "'.$url.'"');
+        }
+
+        if (!array_key_exists('host', $parts)) {
+            throw new InvalidURLException('Provided URL "'.$url.'" doesn\'t contain a hostname');
+        }
+
+        //If credentials are passed in, but we don't want them, raise an exception
+        if (!$options->getSendCredentials() && (array_key_exists('user', $parts) || array_key_exists('pass', $parts))) {
+            throw new InvalidURLException('Credentials passed in but "sendCredentials" is set to false');
+        }
+
+        //First, validate the scheme
+        if (array_key_exists('scheme', $parts)) {
+            $parts['scheme'] = self::validateScheme($parts['scheme'], $options);
+        } else {
+            //Default to http
+            $parts['scheme'] = 'http';
+        }
```

Comment:
You can avoid the `else` and put the default value before the `if`.

Reply:
I can't otherwise it would still be http and never the https, but I used a ternary operator to remove else.